File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    var answer = [];
    let stageCount = Array(N+2).fill(0);
    
    for(let s of stages){
        stageCount[s] += 1;
    }
    
    let totalPlayer = stages.length;
    
    for(let i = 1; i <= N; i++){
        let fail = stageCount[i] / totalPlayer;
        answer.push([i, fail]);
        totalPlayer -= stageCount[i];
    }
    answer.sort((a, b) => b[1] - a[1]);
    
    return answer.map(a => a[0]);
}

File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = list(range(1, N+1))
    fails = {}
    for stage in range(1,N+1):
        # 각 스테이지 별 분모
        botttom = sum(1 for x in stages if x >= stage)
        # 각 스테이지 별 분자
        top = sum(1 for x in stages if x == stage)
        # 각 스테이지 별 실패율
        fails[stage] = top / botttom if botttom != 0 else 0
    
    answer = sorted(answer,key=lambda x: (-fails[x], x))
    
    return answer